iOS/MSMapsModules/MSMapsModules/MSMapGeoJsonParser/MSMapGeoJsonParser.m

```diff
@@ -251,42 +313,135 @@ + (MSGeoposition * _Nullable)parseGeoposition:(NSArray *)coordinates
                       altitudeReferenceSystem:altitudeReferenceSystem];
 }
 
++ (void)verifyPolygonRing:(NSArray *)path error:(NSError **)error {
+  if (path.count < 4) {
+    NSString *array = @"[";
+    for (int i = 0; i < path.count - 1; i++) {
+      array = [array
+          stringByAppendingString:[[MSMapGeoJsonParser
+                                      arrayToString:[path objectAtIndex:i]]
+                                      stringByAppendingString:@", "]];
+    }
+    array = [array
+        stringByAppendingString:[[MSMapGeoJsonParser
+                                    arrayToString:[path
+                                                      objectAtIndex:path.count -
+                                                                    1]]
+                                    stringByAppendingString:@"]"]];
+    *error = [MSMapGeoJsonParser
+        makeGeoJsonError:
+            [@"Polygon ring must have at least 4 positions, and the first and "
+             @"last position must be the same. Instead saw: "
+                stringByAppendingString:array]];
+    return;
+  }
+  NSArray *firstPosition = [path firstObject];
+  NSArray *lastPosition = [path lastObject];
+  if ((firstPosition.count < 3 && lastPosition.count < 3) ||
+      (firstPosition.count > 2 && lastPosition.count > 2)) {
+
+    for (int i = 0; i < MIN(firstPosition.count, lastPosition.count); i++) {
+      if ([firstPosition objectAtIndex:i] != [lastPosition objectAtIndex:i]) {
+        NSString *stringArray = [@"first: "
+            stringByAppendingString:[MSMapGeoJsonParser
+                                        arrayToString:firstPosition]];
+        stringArray = [stringArray
+            stringByAppendingString:
+                [@"  last: ["
+                    stringByAppendingString:[MSMapGeoJsonParser
+                                                arrayToString:lastPosition]]];
+
+        *error = [MSMapGeoJsonParser
+            makeGeoJsonError:[@"First and last coordinate pair of each polygon "
+                              @"ring must be the same. Instead saw: "
+                                 stringByAppendingString:stringArray]];
+        return;
+      }
+    }
+
+  } else {
+    NSString *stringArray =
+        [@"first: " stringByAppendingString:[MSMapGeoJsonParser
+                                                arrayToString:firstPosition]];
+    stringArray = [stringArray
+        stringByAppendingString:
+            [@"  last: ["
+                stringByAppendingString:[MSMapGeoJsonParser
+                                            arrayToString:lastPosition]]];
+
+    *error = [MSMapGeoJsonParser
+        makeGeoJsonError:[@"First and last coordinate pair of each polygon "
+                          @"ring must be the same. Instead saw: "
+                             stringByAppendingString:stringArray]];
+  }
+}
+
++ (MSMapAltitudeReferenceSystem)getAltitudeReferenceSystem:(NSArray *)array
```

Comment:
I considered that approach, but if all points except the last contain an altitude, then once the parser gets to the last point, it would see that all previous points were supposed to be set to 0 surface. It would have to go back and recreate all of the objects with altitude 0 surface. Instead, this scans all the points first to see if any of them lack an altitude coordinate.

Reply:
altitude is a settable property, doesn't look like you have to recreate the object.

Reply:
Resolved in Teams. Making an intermediate object to store data since AltitudeReferenceSystem is read-only for Geopath.